### DISCUSSION TITLE
Is it possible to execute terminal commands on file save?

### DISCUSSION THREAD

**Discussion by vikinatora** [eclipse-theia/theia#13335] [Q&A] [answered]
Hello, I'm exploring the possibility of using Theia as a "sandbox IDE" for a specific library. Theia looks like a good fit, and I've got a question about whether it can run a specific command when I save a file. Here's the flow I'm hoping for:
- Application is opened
- Project template is loaded
- Any file is saved (there may be multiple files in the project).
- Terminal is opened.
- "Execute" command is ran in terminal.
- Results are visible in the terminal.

Assuming the folder and file structure stays the same, I'm guessing I might need an extension for this, but I'm not sure about any limitations on those. If it's not doable, what's the closest thing I can get to what I want with Theia?

**Comment by msujew** [comment, marked as answer]
Hey @vikinatora,

there's little you can't do with Theia, because Theia exposes all of its internal code for overriding purposes via dependency injection. What you're doing is certainly feasible, you just need to hook into the right events and execute the right service methods in your Theia extension.

How these things can be accomplished is another matter. Generally, you would hook into the `onDidSave` event of text editors and then call the `TerminalService` to spawn a new terminal and execute code there. Like with most internal changes to Theia, you'll need to dig into the source code of the framework.

**Reply by vikinatora** [reply]
Thank you! That's the answer I was looking for!
